File: src/features/drawing_tool/DrawingCanvasViewport.js
.pragma library

function finiteNumber(value, fallback) {
    var number = Number(value)
    return Number.isFinite(number) ? number : fallback
}

function boardBounds(viewWidth, viewHeight, zoom, panX, panY) {
    var width = Math.max(0, finiteNumber(viewWidth, 0))
    var height = Math.max(0, finiteNumber(viewHeight, 0))
    var safeZoom = Math.max(0.0001, finiteNumber(zoom, 1.0))
    var safePanX = finiteNumber(panX, 0)
    var safePanY = finiteNumber(panY, 0)
    var board = Math.max(32, Math.min(width, height) - 16) * safeZoom
    return {
        x: Math.round((width - board) / 2 + safePanX),
        y: Math.round((height - board) / 2 + safePanY),
        size: board
    }
}

function canvasToScreenX(bounds, normalizedX) {
    return finiteNumber(bounds && bounds.x, 0) + finiteNumber(normalizedX, 0) * finiteNumber(bounds && bounds.size, 1)
}

function canvasToScreenY(bounds, normalizedY) {
    return finiteNumber(bounds && bounds.y, 0) + finiteNumber(normalizedY, 0) * finiteNumber(bounds && bounds.size, 1)
}

function canvasToScreen(bounds, normalizedX, normalizedY) {
    return {
        x: canvasToScreenX(bounds, normalizedX),
        y: canvasToScreenY(bounds, normalizedY)
    }
}

function screenToCanvas(bounds, screenX, screenY) {
    var size = finiteNumber(bounds && bounds.size, 1)
    if (Math.abs(size) < 0.000001) {
        size = 1
    }
    return {
        x: (finiteNumber(screenX, 0) - finiteNumber(bounds && bounds.x, 0)) / size,
        y: (finiteNumber(screenY, 0) - finiteNumber(bounds && bounds.y, 0)) / size
    }
}

File: src/canvas/DrawingCanvasViewport.cpp
#include "DrawingCanvasViewport.h"

#include <algorithm>
#include <cmath>

namespace drawing_canvas {

BoardBounds boardBounds(double viewWidth, double viewHeight, double zoom, double panX, double panY) {
    const double width = std::max(0.0, finiteNumber(viewWidth, 0.0));
    const double height = std::max(0.0, finiteNumber(viewHeight, 0.0));
    const double safeZoom = std::max(0.0001, finiteNumber(zoom, 1.0));
    const double safePanX = finiteNumber(panX, 0.0);
    const double safePanY = finiteNumber(panY, 0.0);
    const double board = std::max(32.0, std::min(width, height) - 16.0) * safeZoom;
    return {
        std::round((width - board) / 2.0 + safePanX),
        std::round((height - board) / 2.0 + safePanY),
        board
    };
}

double canvasToScreenX(const BoardBounds &bounds, double x) {
    return finiteNumber(bounds.x, 0.0) + finiteNumber(x, 0.0) * finiteNumber(bounds.size, 1.0);
}

double canvasToScreenY(const BoardBounds &bounds, double y) {
    return finiteNumber(bounds.y, 0.0) + finiteNumber(y, 0.0) * finiteNumber(bounds.size, 1.0);
}

ScreenPoint canvasToScreen(const BoardBounds &bounds, const CanvasPoint &point) {
    return {canvasToScreenX(bounds, point.x), canvasToScreenY(bounds, point.y)};
}

CanvasPoint screenToCanvas(const BoardBounds &bounds, const ScreenPoint &point) {
    const double size = std::max(0.000001, finiteNumber(bounds.size, 1.0));
    return {
        (finiteNumber(point.x, 0.0) - finiteNumber(bounds.x, 0.0)) / size,
        (finiteNumber(point.y, 0.0) - finiteNumber(bounds.y, 0.0)) / size
    };
}

BoardBounds boardBoundsFromVariant(const QVariantMap &bounds) {
    return {
        finiteNumber(bounds.value(QStringLiteral("x")), 0.0),
        finiteNumber(bounds.value(QStringLiteral("y")), 0.0),
        std::max(0.000001, finiteNumber(bounds.value(QStringLiteral("size")), 1.0))
    };
}

} // namespace drawing_canvas
